File: 12.py
from aoc import *

DAY = 12

s = get_input(DAY)
# s = get_example(DAY, 6)

import collections
import math

ls = s.splitlines()
n = len(ls)
m = len(ls[0])

def part1():
    ans = 0

    visited = set()
    for i in range(n):
        for j in range(m):
            if (i, j) in visited:
                continue

            area = 0
            perimeter = 0
            q = collections.deque([(i, j)])
            visited.add((i, j))
            while q:
                k, l = q.popleft()
                area += 1
                for dk, dl in DIR4:
                    nk = k + dk
                    nl = l + dl
                    if nk < 0 or nk >= n or nl < 0 or nl >= m or ls[nk][nl] != ls[k][l]:
                        perimeter += 1
                    elif (nk, nl) not in visited:
                        q.append((nk, nl))
                        visited.add((nk, nl))

            ans += area * perimeter

    return ans


def part2():
    ans = 0

    visited = set()
    for i in range(n):
        for j in range(m):
            if (i, j) in visited:
                continue

            area = 0
            sides = 0
            outside_points = set()
            q = collections.deque([(i, j)])
            visited.add((i, j))
            while q:
                k, l = q.popleft()
                area += 1
                for dk, dl in DIR4:
                    nk = k + dk
                    nl = l + dl
                    if nk < 0 or nk >= n or nl < 0 or nl >= m or ls[nk][nl] != ls[k][l]:
                        count = sum(map(lambda d: (nk + d[0], nl + d[1], dk, dl) in outside_points, DIR4))
                        if count == 0:
                            sides += 1
                        elif count > 1:
                            sides -= 1
                        outside_points.add((nk, nl, dk, dl))
                    elif (nk, nl) not in visited:
                        q.append((nk, nl))
                        visited.add((nk, nl))

            ans += area * sides

    return ans


# submit(DAY, part1(), 1)
# submit(DAY, part2(), 2)

File: kotlin/src/12.kt
import java.util.LinkedList
import java.util.Queue
import kotlin.time.measureTimedValue

fun main() {
    val ls = readInput()
    val n = ls.size
    val m = ls[0].length

    fun part1(): Int {
        var ans = 0

        val visited = mutableSetOf<Pair<Int, Int>>()
        for (i in 0 ..< n) {
            for (j in 0 ..< m) {
                val pos = Pair(i, j)
                if (pos in visited)
                    continue

                visited.add(pos)
                val q: Queue<Pair<Int, Int>> = LinkedList()
                q.add(pos)
                var area = 0
                var perimeter = 0
                while (q.isNotEmpty()) {
                    val current = q.remove()
                    area++
                    for (dir in DIR4) {
                        val neighbour = current + dir
                        if (
                            neighbour.first < 0 || neighbour.first >= n ||
                            neighbour.second < 0 || neighbour.second >= m ||
                            ls[neighbour.first][neighbour.second] != ls[current.first][current.second]
                        ) {
                            perimeter++
                        }
                        else if (neighbour !in visited) {
                            visited.add(neighbour)
                            q.add(neighbour)
                        }
                    }
                }
                ans += area * perimeter
            }
        }

        return ans
    }

    fun part2(): Int {
        var ans = 0

        val visited = mutableSetOf<Pair<Int, Int>>()
        for (i in 0 ..< n) {
            for (j in 0 ..< m) {
                val pos = Pair(i, j)
                if (pos in visited)
                    continue

                visited.add(pos)
                val q: Queue<Pair<Int, Int>> = LinkedList()
                q.add(pos)
                var area = 0
                var sides = 0
                val fences = mutableSetOf<Pair<Pair<Int, Int>, Pair<Int, Int>>>()
                while (q.isNotEmpty()) {
                    val current = q.remove()
                    area++
                    for (k in DIR4.indices) {
                        val dir = DIR4[k]
                        val neighbour = current + dir
                        if (
                            neighbour.first < 0 || neighbour.first >= n ||
                            neighbour.second < 0 || neighbour.second >= m ||
                            ls[neighbour.first][neighbour.second] != ls[current.first][current.second]
                        ) {
                            when (listOf(DIR4[(k - 1 + DIR4.size) % DIR4.size], DIR4[(k + 1) % DIR4.size]).count { Pair(neighbour + it, dir) in fences }) {
                                0 -> sides++
                                2 -> sides--
                            }
                            fences.add(Pair(neighbour, dir))
                        }
                        else if (neighbour !in visited) {
                            visited.add(neighbour)
                            q.add(neighbour)
                        }
                    }
                }
                ans += area * sides
            }
        }

        return ans
    }

    println(measureTimedValue { part1() }.let { "${it.value} (${it.duration})" })
    println(measureTimedValue { part2() }.let { "${it.value} (${it.duration})" })
}
